pwiz_tools/Skyline/Util/PanoramaUtil.cs

```diff
@@ -151,21 +129,47 @@ public static UserState ValidateServerAndUser(ref Uri serverUri, string username
                     }
                 }
 
-                return UserState.unknown;
+                return new UserState(UserStateEnum.unknown, ex.Message, GetEnsureLoginUri(pServer));
             }
         }
 
         private static UserState EnsureLogin(PanoramaServer pServer)
         {
-            var requestUri = new Uri(pServer.ServerUri, ENSURE_LOGIN_PATH);
+            var requestUri = GetEnsureLoginUri(pServer);
             var request = (HttpWebRequest) WebRequest.Create(requestUri);
             request.Headers.Add(HttpRequestHeader.Authorization,
                 Server.GetBasicAuthHeader(pServer.Username, pServer.Password));
             try
             {
                 using (var response = (HttpWebResponse) request.GetResponse())
                 {
-                    return response.StatusCode == HttpStatusCode.OK ? UserState.valid : UserState.unknown;
+                    if (response.StatusCode != HttpStatusCode.OK)
+                    {
+                        return new UserState(UserStateEnum.nonvalid,

```

Comment:
Consider how this would look if you simply threw an exception at this point with all of the state information on it.

Reply:
I will do this in a separate commit.  It requires some work.